ref(traffic_sim): AJD-696 clean up metics, provide out_of_range

dmoszynski opened the pull request:
## Types of PR
- [ ] New Features
- [x] Upgrade of existing features
- [ ] Bugfix

## Link to the issue
https://tier4.atlassian.net/browse/AJD-696

## Description

- Removed all metrics - `metrics_manager_` no longer exists in `traffic_simulator::API`.
- Based on `metric::OutOfRangeMetric`, a new `job` `OUT_OF_RANGE` [was developed](https://github.com/tier4/scenario_simulator_v2/blob/e504f6fad626a32e050f5dc4aa1f15d0a027f2ec/simulation/traffic_simulator/src/entity/entity_base.cpp#L726-L766) - it is forwarded `API->EntityManager->EntityBase`
- Job `OUT_OF_RANGE` is activated in [activatePerformanceAssertion](https://github.com/tier4/scenario_simulator_v2/blob/e504f6fad626a32e050f5dc4aa1f15d0a027f2ec/openscenario/openscenario_interpreter/include/openscenario_interpreter/simulator_core.hpp#L391-L394) - as it was with the metric.
- Job has the same (as the metric) fixed limit values, passed at the time of activation.
- Job is activated for `Vehicles` (including `Ego`) - it is added to their `job_list_`.  
- Everything related to metrics (cpp mock, documentation, etc.) has been removed from the whole code.

## How to review this PR.

## Others

hakuturu583 commented on openscenario/openscenario_interpreter/include/openscenario_interpreter/simulator_core.hpp:
```diff
@@ -380,26 +387,6 @@ class SimulatorCore
   class NonStandardOperation : private CoordinateSystemConversion
   {
   protected:
-    template <typename Performance, typename Properties>
-    static auto activatePerformanceAssertion(
-      const std::string & entity_ref, const Performance & performance,
-      const Properties & properties)
-    {
-      core->addMetric<metrics::OutOfRangeMetric>(entity_ref + "-out-of-range", [&]() {
-        metrics::OutOfRangeMetric::Config configuration;
-        configuration.target_entity = entity_ref;
-        configuration.min_velocity = -performance.max_speed;
-        configuration.max_velocity = +performance.max_speed;
-        configuration.min_acceleration = -performance.max_deceleration;
-        configuration.max_acceleration = +performance.max_acceleration;
-        configuration.min_jerk = properties.template get<Double>("minJerk", Double::lowest());
-        configuration.max_jerk = properties.template get<Double>("maxJerk", Double::max());
-        configuration.jerk_topic =
-          "/planning/scenario_planning/motion_velocity_optimizer/closest_jerk";
-        return configuration;
-      }());
-    }
```
Why do you completely remove this section?

hakuturu583 replied:
Yes, this change does not affect the OpenSCENARIO syntax, so inside the interpreter, the part that used to call the api::addMetric function is simply changed to a function such as api::checkOutOfRange.
The change to the openscenario_interpretor package should be much smaller.

hakuturu583 replied:
It is advisable to REVERT this part and change the member functions of the API class that are called inside the activatePerformanceAssertion function.

dmoszynski replied:
This change is related to a [comment here](https://github.com/tier4/scenario_simulator_v2/pull/979#issuecomment-1482449490) - please read it first.
As I understand it, you suggest adding the job to the `activatePerformanceAssertion(...)` - instead of [in the constructor](https://github.com/tier4/scenario_simulator_v2/blob/cf575c0fb9e9bbb6f62453fb4aa6fdf552b0eb4a/simulation/traffic_simulator/src/entity/vehicle_entity.cpp#L41). Do you think it is necessary?

hakuturu583 replied:
Yes, I have an activatePerformanceAssertion(...) function and do not see the need to change it

hakuturu583 commented on openscenario/openscenario_interpreter/src/syntax/add_entity_action.cpp:
```diff
@@ -79,11 +79,6 @@ try {
 
       TeleportAction::teleport(entity_ref, position);
       AssignControllerAction(entity.as<ScenarioObject>().object_controller)(entity_ref);
-      activatePerformanceAssertion(
-        entity_ref, vehicle.performance,
-        entity.as<ScenarioObject>().object_controller.is<Controller>()
-          ? entity.as<ScenarioObject>().object_controller.as<Controller>().properties
-          : Properties());
```
Why do you completely remove this section?

dmoszynski replied:
Because this method has become empty: [comment](https://github.com/tier4/scenario_simulator_v2/pull/979/files#r1146078172)

hakuturu583 commented on openscenario/openscenario_interpreter/src/syntax/dynamic_constraints.cpp:
```diff
@@ -37,7 +38,8 @@ DynamicConstraints::DynamicConstraints(const pugi::xml_node & node, Scope & scop
     readAttribute<Double>("maxDecelerationRate", node, scope, Double::infinity())),
-  max_speed(readAttribute<Double>("maxSpeed", node, scope, Double::infinity()))
+  max_speed(readAttribute<Double>("maxSpeed", node, scope, Double::infinity())),
+  max_jerk(readAttribute<Double>("maxJerk", node, scope, Double::infinity()))
```
maxJerk field does not exist in [DynamicConstraints.](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/DynamicConstraints.html)
Please remove it.

hakuturu583 commented on openscenario/openscenario_interpreter/src/syntax/dynamic_constraints.cpp:
```diff
@@ -49,6 +51,7 @@ DynamicConstraints::operator traffic_simulator_msgs::msg::DynamicConstraints() c
   dynamic_constraints.max_deceleration = max_deceleration;
   dynamic_constraints.max_deceleration_rate = max_deceleration_rate;
   dynamic_constraints.max_speed = max_speed;
+  dynamic_constraints.max_jerk = max_jerk;
```
maxJerk field does not exist in [DynamicConstraints.](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/DynamicConstraints.html)
Please remove it.

hakuturu583 commented on openscenario/openscenario_interpreter/include/openscenario_interpreter/syntax/dynamic_constraints.hpp:
```diff
@@ -52,12 +53,14 @@ struct DynamicConstraints
 
   const Double max_speed;
 
+  const Double max_jerk;
```
maxJerk field does not exist in [DynamicConstraints.](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/DynamicConstraints.html)
Please remove it.

hakuturu583 commented on openscenario/openscenario_interpreter/include/openscenario_interpreter/syntax/dynamic_constraints.hpp:
```diff
@@ -52,12 +53,14 @@ struct DynamicConstraints
     const Double max_deceleration = Double::infinity(),
     const Double max_deceleration_rate = Double::infinity(),
-    const Double max_speed = Double::infinity());
+    const Double max_speed = Double::infinity(), const Double max_jerk = Double::infinity());
```
maxJerk field does not exist in [DynamicConstraints.](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/DynamicConstraints.html)
Please remove it.

hakuturu583 commented on openscenario/openscenario_interpreter/src/syntax/performance.cpp:
```diff
@@ -21,12 +21,13 @@ inline namespace syntax
 Performance::Performance(
   const Double max_acceleration, const Double max_acceleration_rate, const Double max_deceleration,
-  const Double max_deceleration_rate, const Double max_speed)
+  const Double max_deceleration_rate, const Double max_speed, const Double max_jerk)
```
maxJerk field does not exist in [Performance.](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/Performance.html)
Please remove it.

hakuturu583 commented on openscenario/openscenario_interpreter/src/syntax/performance.cpp:
```diff
@@ -21,12 +21,13 @@ inline namespace syntax
   max_deceleration_rate(max_deceleration_rate),
-  max_speed(max_speed)
+  max_speed(max_speed),
+  max_jerk(max_jerk)
```
maxJerk field does not exist in [Performance.](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/Performance.html)
Please remove it.

hakuturu583 commented on openscenario/openscenario_interpreter/src/syntax/performance.cpp:
```diff
@@ -37,7 +38,8 @@ Performance::Performance(const pugi::xml_node & node, Scope & scope)
     readAttribute<Double>("maxDecelerationRate", node, scope, Double::infinity())),
-  max_speed(readAttribute<Double>("maxSpeed", node, scope, Double::infinity()))
+  max_speed(readAttribute<Double>("maxSpeed", node, scope, Double::infinity())),
+  max_jerk(readAttribute<Double>("maxJerk", node, scope, Double::infinity()))
```
maxJerk field does not exist in [Performance.](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/Performance.html)
Please remove it.

hakuturu583 commented on openscenario/openscenario_interpreter/src/syntax/performance.cpp:
```diff
@@ -49,6 +51,7 @@ Performance::operator traffic_simulator_msgs::msg::Performance() const
   performance.max_deceleration = max_deceleration;
   performance.max_deceleration_rate = max_deceleration_rate;
   performance.max_speed = max_speed;
+  performance.max_jerk = max_jerk;
```
maxJerk field does not exist in [Performance.](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/Performance.html)
Please remove it.

hakuturu583 commented on simulation/traffic_simulator/src/entity/ego_entity.cpp:
```diff
@@ -161,6 +164,9 @@ EgoEntity::EgoEntity(
   vehicle_model_type_(getVehicleModelType()),
   vehicle_model_ptr_(makeSimulationModel(vehicle_model_type_, step_time, parameters))
 {
+  // deactivate non-Ego jobs
+  // TODO: this job should be activate but needs data obtained from Autoware in BehaviorParameter
+  job_list_.inactivate(traffic_simulator::job::Type::OUT_OF_RANGE_DYNAMIC_CONSTRAINS);
```
Why inactivate?

dmoszynski replied:
[Here](https://tier4.atlassian.net/browse/AJD-696?focusedCommentId=117435) i commented:

> Should Performance (from the scenario) be taken into account when setting DynamicConstraints for Ego? (last column in the table)
> 
> I encountered a problem with [fixed](https://github.com/tier4/scenario_simulator_v2/blob/c425a60edc321fcfd56d4404ec42f0a64865d3e4/simulation/traffic_simulator/src/entity/ego_entity.cpp#L178-L188) max_acceleration==0 for Ego so I made some changes to ensure the effect of the Performance values. ([commit](https://github.com/tier4/scenario_simulator_v2/pull/979/commits/54f5abce604da9d679e207a82fe99671a4fb8964))

You answered:

> Why do you remove auto EgoEntity::getBehaviorParameter() const -> traffic_simulator_msgs::msg::BehaviorParameter , [...]?
> I think default value should be const.

I answered:

> I'm not sure why there were assignments like max acceleration=0 in EgoEntity::getBehaviorParameter() - without commenting out [these 2 lines](https://github.com/tier4/scenario_simulator_v2/blob/c425a60edc321fcfd56d4404ec42f0a64865d3e4/simulation/traffic_simulator/src/entity/ego_entity.cpp#L185-L186), out_of_range would not allow Ego to move.
> 
> However, this method still returned msgs::BehaviourParameter - not taking Performance into account. That's why I’ve removed this method and the definition from [VehicleEntity](https://github.com/tier4/scenario_simulator_v2/blob/54f5abce604da9d679e207a82fe99671a4fb8964/simulation/traffic_simulator/src/entity/vehicle_entity.cpp#L74-L77) is used.

You answered:

> This line [( entity/ego_entity.cpp#L181-L187)](https://github.com/tier4/scenario_simulator_v2/blob/f9357649bce1b2990574086398061d3afe2cba37/simulation/traffic_simulator/src/entity/ego_entity.cpp#L181-L187) shows that the part is still WIP and should not be changed without checking and considering the circumstances of why this is happening.
> I don't think it is a good implementation to use a base type function without inputting data obtained from Autoware in BehaviorParameter in the first place.

I answered:

> I see, thank you. I will undo the changes from this [commit](https://github.com/tier4/scenario_simulator_v2/commit/8215ade14ca5ad3cfa3c025f2d2fb5797884f150). 
> However, in this case, the job out_of_range for Ego can’t be run - I understand that it should only be run for non-Ego Vehicles?

So I've reverted `EgoEntity::getBehaviorParameter()` to the previous version and deactivated the `out_of_range` job for Ego - because there is a fixed value of `max_acceleration=0`.

hakuturu583 commented on simulation/traffic_simulator/src/job/job_list.cpp:
```diff
@@ -31,6 +31,12 @@ void JobList::append(
   list_.emplace_back(Job(func_on_update, func_on_cleanup, type, exclusive, event));
 }
 
+void JobList::inactivate(const job::Type type)
+{
+  for (auto & job : list_)
+    if (job.type == type) job.inactivate();
+}
+
```
I think this function is unnecessary.

dmoszynski commented on openscenario/openscenario_interpreter/include/openscenario_interpreter/syntax/performance.hpp:
```diff
@@ -47,12 +48,14 @@ struct Performance
     const Double max_deceleration = Double::infinity(),
     const Double max_deceleration_rate = Double::infinity(),
-    const Double max_speed = Double::infinity());
+    const Double max_speed = Double::infinity(), const Double max_jerk = Double::infinity());
```
When I was developing `out_of_range` I needed a value limits for jerk. 
I thought adding `maxJerk` to [DynamicConstraints.msg](https://github.com/tier4/scenario_simulator_v2/blob/d43f5dd9d7f645ab4ea1c649fc74301e84c0cf71/simulation/traffic_simulator_msgs/msg/DynamicConstraints.msg#L6) and [Performance.msg](https://github.com/tier4/scenario_simulator_v2/blob/d43f5dd9d7f645ab4ea1c649fc74301e84c0cf71/simulation/traffic_simulator_msgs/msg/Performance.msg) was necessary to keep this simply and ensure clean code.
That's why I adapted the rest of the code ([commit](https://github.com/tier4/scenario_simulator_v2/pull/979/commits/6c0b8b664a4e650c9b5039d4d9f924e286470624)).
I wasn't aware that  [Performance.msg](https://github.com/tier4/scenario_simulator_v2/blob/d43f5dd9d7f645ab4ea1c649fc74301e84c0cf71/simulation/traffic_simulator_msgs/msg/Performance.msg) had to be compatible with [Performance](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/Performance.html).
Of course, I will revert all `maxJerk/max_jerk` changes,
Please tell me how can i provide `jerk` limits for `out_of_range` - because in that case i have no idea.

hakuturu583 replied:
I suggest to add functions like this 

```
API::checkOutofRange(const std::string & entity_name)
```

When the API::checkOutOfRange function calls, the entity append checkOutOfRnage job.
If you implement like this, you do not pass jerk limit on constructor in entity class.

dmoszynski commented on openscenario/openscenario_interpreter/src/syntax/dynamic_constraints.cpp:
```diff
@@ -21,12 +21,13 @@ inline namespace syntax
 DynamicConstraints::DynamicConstraints(
   const Double max_acceleration, const Double max_acceleration_rate, const Double max_deceleration,
-  const Double max_deceleration_rate, const Double max_speed)
+  const Double max_deceleration_rate, const Double max_speed, const Double max_jerk)
```
When I was developing `out_of_range` I needed a value limits for jerk.
I thought adding `maxJerk` to [DynamicConstraints.msg](https://github.com/tier4/scenario_simulator_v2/blob/d43f5dd9d7f645ab4ea1c649fc74301e84c0cf71/simulation/traffic_simulator_msgs/msg/DynamicConstraints.msg#L6) and [Performance.msg](https://github.com/tier4/scenario_simulator_v2/blob/d43f5dd9d7f645ab4ea1c649fc74301e84c0cf71/simulation/traffic_simulator_msgs/msg/Performance.msg) was necessary to keep this simply and ensure clean code.
That's why I adapted the rest of the code ([commit](https://github.com/tier4/scenario_simulator_v2/pull/979/commits/6c0b8b664a4e650c9b5039d4d9f924e286470624)).
I wasn't aware that [DynamicConstraints.msg](https://github.com/tier4/scenario_simulator_v2/blob/d43f5dd9d7f645ab4ea1c649fc74301e84c0cf71/simulation/traffic_simulator_msgs/msg/DynamicConstraints.msg#L6)  had to be compatible with [DynamicConstraints](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/DynamicConstraints.html)
Of course, I will revert all `maxJerk/max_jerk` changes,
Please tell me how can i provide `jerk` limits for `out_of_range` - because in that case i have no idea.

dmoszynski commented:
@hakuturu583 
I've reverted all changes related to `maxJerk/max_jerk`.
I've added `max_jerk` as a variable in the class [EntityBase](https://github.com/tier4/scenario_simulator_v2/blob/9c047af8437bff04330b2f0dbef5afe913c5b2a4/simulation/traffic_simulator/include/traffic_simulator/entity/entity_base.hpp#L251).
My change related to [setJerkLimit()](https://github.com/tier4/scenario_simulator_v2/blob/9c047af8437bff04330b2f0dbef5afe913c5b2a4/simulation/traffic_simulator/src/entity/entity_base.cpp#L772-L777), ensured that [max_jerk](https://github.com/tier4/scenario_simulator_v2/blob/9c047af8437bff04330b2f0dbef5afe913c5b2a4/simulation/traffic_simulator/include/traffic_simulator/entity/entity_base.hpp#L251) is set based on `maxJerk` in scenarios.
I've tested this by adding `maxJerk==0.1` in the scenario with `Bicycle0` - the exception is thrown correctly.
![Screenshot from 2023-03-23 13-43-09](https://user-images.githubusercontent.com/121798334/227207362-cd316d2a-f4f0-4c96-80e5-4d85ffb97804.png)
![Screenshot from 2023-03-23 13-29-14](https://user-images.githubusercontent.com/121798334/227207346-3b6019fb-7068-4312-a101-4a7f2d249854.png)
Again, I apologize for the earlier changes with `maxJerk` in DynamicConstraints, I wasn't aware that `maxJerk` couldn't be there.
Please let me know if this solution is correct.

hakuturu583 commented on simulation/traffic_simulator/include/traffic_simulator/job/job_list.hpp:
```diff
@@ -30,6 +30,7 @@ class JobList
     const std::function<void()> & func_on_cleanup, job::Type type, bool exclusive,
     const job::Event event);
   void update(const double step_time, const job::Event event);
+  void inactivate(const job::Type type);
```
I think inactive member function is unncessary.

hakuturu583 commented on simulation/traffic_simulator/src/entity/ego_entity.cpp:
```diff
@@ -161,6 +161,8 @@ EgoEntity::EgoEntity(
   vehicle_model_ptr_(makeSimulationModel(vehicle_model_type_, step_time, parameters))
 {
+  // TODO: this job should be activate but needs data obtained from Autoware in BehaviorParameter
+  job_list_.inactivate(traffic_simulator::job::Type::OUT_OF_RANGE_DYNAMIC_CONSTRAINS);
```
Out of range check is necessary for EgoEntity.

dmoszynski replied:
@hakuturu583 
I also believe that `out_of_range` should be activated for Ego. However, as described [here](https://github.com/tier4/scenario_simulator_v2/pull/979#discussion_r1145919231), because of [this](https://github.com/tier4/scenario_simulator_v2/blob/cf575c0fb9e9bbb6f62453fb4aa6fdf552b0eb4a/simulation/traffic_simulator/src/entity/ego_entity.cpp#L180-L190) method cannot be:
```
auto EgoEntity::getBehaviorParameter() const -> traffic_simulator_msgs::msg::BehaviorParameter
{
  traffic_simulator_msgs::msg::BehaviorParameter parameter;
  /**
   * @brief TODO, Input values get from autoware.
   */
  parameter.see_around = true;
  parameter.dynamic_constraints.max_acceleration = 0;
  parameter.dynamic_constraints.max_deceleration = 0;
  return parameter;
}
```
I mentioned it in a comment [here](https://tier4.atlassian.net/browse/AJD-696?focusedCommentId=117533) and suggested changes ([commit](https://github.com/tier4/scenario_simulator_v2/pull/979/commits/54f5abce604da9d679e207a82fe99671a4fb8964)):

> I'm not sure why there were assignments like max acceleration=0 in EgoEntity::getBehaviorParameter() - without commenting out [these 2 lines](https://github.com/tier4/scenario_simulator_v2/blob/c425a60edc321fcfd56d4404ec42f0a64865d3e4/simulation/traffic_simulator/src/entity/ego_entity.cpp#L185-L186), out_of_range would not allow Ego to move.
> 
> However, this method still returned msgs::BehaviourParameter - not taking Performance into account. That's why I’ve removed this method and the definition from [VehicleEntity](https://github.com/tier4/scenario_simulator_v2/blob/54f5abce604da9d679e207a82fe99671a4fb8964/simulation/traffic_simulator/src/entity/vehicle_entity.cpp#L74-L77) is used.
> 
> I'm not sure if Performance should be taken into account in these cases, thats why I asked. :slight_smile:

In summary: 
In order to run `out_of_range` for Ego it is necessary to change the definition of `EgoEntity::getBehaviorParameter()`. 
For this reason I removed the definitions of `EgoEntity::getBehaviorParameter()` and `EgoEntity::setBehaviorParameter()` - the definition from [VehicleEntity](https://github.com/tier4/scenario_simulator_v2/blob/54f5abce604da9d679e207a82fe99671a4fb8964/simulation/traffic_simulator/src/entity/vehicle_entity.cpp#L74-L77) was used. 
This allowed not only to activate the job for Ego, but also ensures that Performance limits were taken into account.


However, you answered:
> This line [( entity/ego_entity.cpp#L181-L187)](https://github.com/tier4/scenario_simulator_v2/blob/f9357649bce1b2990574086398061d3afe2cba37/simulation/traffic_simulator/src/entity/ego_entity.cpp#L181-L187) shows that the part is still WIP and should not be changed without checking and considering the circumstances of why this is happening.
> I don't think it is a good implementation to use a base type function without inputting data obtained from Autoware in BehaviorParameter in the first place.

So I thought that since [these lines](https://github.com/tier4/scenario_simulator_v2/blob/c425a60edc321fcfd56d4404ec42f0a64865d3e4/simulation/traffic_simulator/src/entity/ego_entity.cpp#L185-L186) _should not be changed_, reverting changes and deactivating `out_of_range` for Ego is the only solution. 

If indeed  `EgoEntity::getBehaviorParameter()` cannot be modified and this check is necessary for `EgoEntity`:
I suggest to redefine `addOutOfRangeJob()` in `EgoEntity` and there provide const limits from raw `traffic_simulator_msgs::msg::DynamicConstraints` message:
```
void EgoEntity::addOutOfRangeJob()
{
  /**
   * @brief This value was determined heuristically rather than for
   * technical reasons.
   */
  constexpr double tolerance = 0.01;
  job_list_.append(
    /**
     * @brief Checking if the values of velocity, acceleration and jerk are within the acceptable
     * range
     */
    [this, tolerance](double) {
      auto velocity_ = status_.action_status.twist.linear.x;
      auto accel_ = status_.action_status.accel.linear.x;
      auto jerk_ = status_.action_status.linear_jerk;
      auto d_c = traffic_simulator_msgs::msg::DynamicConstraints;
      if (std::abs(velocity_) - d_c.max_speed > tolerance)
        THROW_SPECIFICATION_VIOLATION(
          "Entity: ", name, " - current velocity (which is ", velocity_,
          ") is out of range (which is [", -d_c.max_speed, ", ", d_c.max_speed, "])");

      if (accel_ - d_c.max_acceleration > tolerance || -accel_ - d_c.max_deceleration > tolerance)
        THROW_SPECIFICATION_VIOLATION(
          "Entity: ", name, " - current acceleration (which is ", accel_,
          ") is out of range (which is [", -d_c.max_deceleration, ", ", d_c.max_acceleration, "])");

      if (std::abs(jerk_) - max_jerk_ > tolerance)
        THROW_SPECIFICATION_VIOLATION(
          "Entity: ", name, " - current jerk (which is ", jerk_, ") is out of range (which is [",
          -max_jerk_, ", ", max_jerk_, "])");
      return false;
    },
    /**
     * @brief This job is always ACTIVE
     */
    [this]() {}, job::Type::OUT_OF_RANGE_DYNAMIC_CONSTRAINS, true, job::Event::POST_UPDATE);
}
```

hakuturu583 replied:
OutOfRangeMetric is a metric to verify if the velocity, etc. is within the specified range, and we have confirmed that there is no need to follow if DynamicConstriaints, etc. are subsequently updated.
As with the current implementation of openscenario_interpretor, after calling API::spawn function, API::checkOutofRange(const std::string & entity_name, speed_limit, jerk_limit, etc...) function to initiate the check.

Translated with www.DeepL.com/Translator (free version)

hakuturu583 commented:
I add some comment, how do you think about this implentation?

https://github.com/tier4/scenario_simulator_v2/pull/979#discussion_r1147112715

dmoszynski commented:
@hakuturu583 

> 
> I suggest to add functions like this
> 
> `API::checkOutofRange(const std::string & entity_name, speed_limit, jerk_limit, etc...)`
> 
> When the API::checkOutOfRange function calls, the entity append checkOutOfRnage job.
> If you implement like this, you do not pass jerk limit on constructor in entity class.
> 
I believe that such a solution is not necessary, because now I also do not pass jerk limit (or any limit)  on constructor in entity class.
Moreover, developed job `out_of_range` works at the limits taken from `getBehaviorParameter().dynamic_constraints`.
Thanks to this, if the `DynamicConstraints` for some Entity change - `out_of_range` will take into account the updated values.
If I understand correctly, limits in `DynamicConstraints` can be changed by methods:
 - [setDynamicConstrains()](https://github.com/tier4/scenario_simulator_v2/blob/cf575c0fb9e9bbb6f62453fb4aa6fdf552b0eb4a/simulation/traffic_simulator/src/entity/entity_base.cpp#L681-L687)
 - [setVelocityLimit()](https://github.com/tier4/scenario_simulator_v2/blob/cf575c0fb9e9bbb6f62453fb4aa6fdf552b0eb4a/simulation/traffic_simulator/src/entity/entity_base.cpp#L765-L770)
- [setJerkLimit()](https://github.com/tier4/scenario_simulator_v2/blob/cf575c0fb9e9bbb6f62453fb4aa6fdf552b0eb4a/simulation/traffic_simulator/src/entity/entity_base.cpp#L772-L777)
- [setAccelerationLimit()](https://github.com/tier4/scenario_simulator_v2/blob/9c047af8437bff04330b2f0dbef5afe913c5b2a4/simulation/traffic_simulator/src/entity/vehicle_entity.cpp#L243-L251)
- [setDecelerationLimit()](https://github.com/tier4/scenario_simulator_v2/blob/9c047af8437bff04330b2f0dbef5afe913c5b2a4/simulation/traffic_simulator/src/entity/vehicle_entity.cpp#L263-L271)
- [resetDynamicConstraints()](https://github.com/tier4/scenario_simulator_v2/blob/cf575c0fb9e9bbb6f62453fb4aa6fdf552b0eb4a/simulation/traffic_simulator/src/entity/entity_base.cpp#L308-L311)

Which are used for example in `EntityBase::requestSpeedChangeWithConstantAcceleration(...)` ([link](https://github.com/tier4/scenario_simulator_v2/blob/9c047af8437bff04330b2f0dbef5afe913c5b2a4/simulation/traffic_simulator/src/entity/entity_base.cpp#L359-L408))
So if the `DynamicConstraints` are changed [here](https://github.com/tier4/scenario_simulator_v2/blob/9c047af8437bff04330b2f0dbef5afe913c5b2a4/simulation/traffic_simulator/src/entity/entity_base.cpp#L375) I believe that setting fixed limits when creating a job may not be a good solution.
I believe that out_of_range should always have the currently valid limits 
(based on the current values in `getBehaviorParameter().dynamic_constraints`)

hakuturu583 commented:
> I believe that such a solution is not necessary, because now I also do not pass jerk limit (or any limit) on constructor in entity class.
Moreover, developed job out_of_range works at the limits taken from getBehaviorParameter().dynamic_constraints.
Thanks to this, if the DynamicConstraints for some Entity change - out_of_range will take into account the updated values.
If I understand correctly, limits in DynamicConstraints can be changed by methods:

OutOfRange Metric is a function to check that the current implementation in the master branch does not take into account DynamicConstraints updates and that it does not produce more than the jerk specified in the scenario.
Therefore, it is not necessary to consider updates.

dmoszynski commented:
@hakuturu583 
I've developed another solution to activate the `OUT_OF_RANGE` job.
I've updated the PR description - please take a look.
In summary: 
Now the job is activated in  [activatePerformanceAssertion](https://github.com/tier4/scenario_simulator_v2/blob/e504f6fad626a32e050f5dc4aa1f15d0a027f2ec/openscenario/openscenario_interpreter/include/openscenario_interpreter/simulator_core.hpp#L391-L394) - as it was with the metric.
Value limits are passed in the activation method - the limits are fixed, aren't updated when `DynamicConstrains` changes in Entities.
I believe that the current solution introduces much less changes and the developed job is fully equivalent to the metric.

hakuturu583 commented on simulation/traffic_simulator/src/entity/entity_base.cpp:
```diff
@@ -721,8 +723,52 @@ void EntityBase::setTrafficLightManager(
+
 auto EntityBase::setVelocityLimit(double) -> void {}
 
+auto EntityBase::setJerkLimit(double) -> void {}
```
This function does not appear to do anything.
What is the purpose of this function?

dmoszynski replied:
This is related to an earlier solution where the `max_jerk` was set by [this call](https://github.com/tier4/scenario_simulator_v2/blob/e504f6fad626a32e050f5dc4aa1f15d0a027f2ec/openscenario/openscenario_interpreter/include/openscenario_interpreter/simulator_core.hpp#L236-L238).
As you can see now `max_jerk` is passed in [activation](https://github.com/tier4/scenario_simulator_v2/blob/e504f6fad626a32e050f5dc4aa1f15d0a027f2ec/openscenario/openscenario_interpreter/include/openscenario_interpreter/simulator_core.hpp#L391-L394) so now `EntityBase::setJerkLimit()` do nothing - but I've decided to leave it as I assume it will be used in the future.

I asked about it [here](https://tier4.atlassian.net/browse/AJD-696?focusedCommentId=117435):
> If I understand correctly, it should have an effect and calling these methods [here](https://github.com/tier4/scenario_simulator_v2/blob/6c0b8b664a4e650c9b5039d4d9f924e286470624/openscenario/openscenario_interpreter/include/openscenario_interpreter/simulator_core.hpp#L236-L242) should do [it](https://github.com/tier4/scenario_simulator_v2/blob/8215ade14ca5ad3cfa3c025f2d2fb5797884f150/simulation/traffic_simulator/src/entity/entity_base.cpp#L766-L778). ([commit](https://github.com/tier4/scenario_simulator_v2/commit/8215ade14ca5ad3cfa3c025f2d2fb5797884f150))

You answered:

> Yes, it should affect.
> Your commit is looks good to me, thanks.

Since the definition of `EntityBase::setVelocityLimit` was empty before my earlier changes, I left `EntityBase::setJerkLimit ` empty as well.

hakuturu583 replied:
I think this function it was necessary, but unnecessary now.
Can you remove this function?

dmoszynski replied:
Yes of course - I've removed [setJerkLimit](https://github.com/tier4/scenario_simulator_v2/pull/979/commits/034d844104e3b78416cf83de0ff2e061e648c2e9)

hakuturu583 replied:
Thanks!
